src/AppInstallerCLICore/Workflows/PinFlow.cpp

```diff
@@ -0,0 +1,247 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+#include "pch.h"
+#include "Resources.h"
+#include "PinFlow.h"
+#include "TableOutput.h"
+#include "Microsoft/PinningIndex.h"
+#include "Microsoft/SQLiteStorageBase.h"
+#include "winget/RepositorySearch.h"
+
+using namespace AppInstaller::Repository;
+using namespace AppInstaller::Repository::Microsoft;
+
+namespace AppInstaller::CLI::Workflow
+{
+    namespace
+    {
+        // Creates a Pin appropriate for the context based on the arguments provided
+        Pinning::Pin CreatePin(Execution::Context& context, std::string_view packageId, std::string_view sourceId, const std::string& installedVersion)
+        {
+            if (context.Args.Contains(Execution::Args::Type::GatedVersion))
+            {
+                return Pinning::Pin::CreateGatingPin({ packageId, sourceId }, context.Args.GetArg(Execution::Args::Type::GatedVersion));
+            }
+            else if (context.Args.Contains(Execution::Args::Type::BlockingPin))
+            {
+                return Pinning::Pin::CreateBlockingPin({ packageId, sourceId }, { installedVersion });
+            }
+            else
+            {
+                return Pinning::Pin::CreatePinningPin({ packageId, sourceId }, { installedVersion });
+            }
+        }
+    }
+
+    void OpenPinningIndex(Execution::Context& context)
+    {
+        auto indexPath = Runtime::GetPathTo(Runtime::PathName::LocalState) / "pinning.db";
+
+        AICLI_LOG(CLI, Info, << "Opening pinning index");
+
+        auto pinningIndex = std::filesystem::exists(indexPath) ?
+            PinningIndex::Open(indexPath.u8string(), SQLiteStorageBase::OpenDisposition::ReadWrite) :
+            PinningIndex::CreateNew(indexPath.u8string());

```

Comment:
This needs to be in the RepositoryCore project so that we can search the pins when readding a package. I'll move it in a future PR when I add that.